lls/arp.c

```diff
@@ -118,21 +123,34 @@ xmit_arp_req(struct gatekeeper_if *iface, const uint8_t *ip_be,
 
 int
 process_arp(struct lls_config *lls_conf, struct gatekeeper_if *iface,
-	uint16_t tx_queue, struct rte_mbuf *buf, struct ether_hdr *eth_hdr)
+	uint16_t tx_queue, struct rte_mbuf *buf, struct ether_hdr *eth_hdr,
+	struct arp_hdr *arp_hdr)
 {
 	struct lls_mod_req mod_req;
-	struct arp_hdr *arp_hdr;
 	uint16_t pkt_len = rte_pktmbuf_data_len(buf);
+	/* pkt_in_skip_l2() already called by LLS. */
+	size_t l2_len = pkt_in_l2_hdr_len(buf);
 
-	if (pkt_len < sizeof(*eth_hdr) + sizeof(*arp_hdr)) {
+	if (pkt_len < l2_len + sizeof(*arp_hdr)) {
 		RTE_LOG(ERR, GATEKEEPER, "lls: %s interface received ARP packet of size %hu bytes, but it should be at least %zu bytes\n",
 			iface->name, pkt_len,
-			sizeof(*eth_hdr) + sizeof(*arp_hdr));
+			l2_len + sizeof(*arp_hdr));
 		return -1;
 	}
 
-	arp_hdr = rte_pktmbuf_mtod_offset(buf, struct arp_hdr *,
-		sizeof(struct ether_hdr));
+	/*
+	 * Check VLAN tag. If it's incorrect, issue a warning
+	 * and correct it in case we reuse this packet.
+	 */
+	if (iface->vlan_insert) {
+		struct vlan_hdr *vlan_hdr = (struct vlan_hdr *)&eth_hdr[1];
```

Comment:
If `iface->vlan_insert` is true, what guarantees that the packet has a VLAN header? The other way around, if If `iface->vlan_insert` is false, what guarantees that the packet does not have a VLAN header? You may need to change the code to enforces these guarantees, or just write a comment explaining how they are currently enforced.

The same goes for `process_nd_neigh_solicitation()`.